Deprecation warnings break the high-level client in strict mode

When the high-level client uses a low-level client that fails on deprecation warnings, that breaks the high-level client quite badly. The low-level client throws a `ResponseException` which is caught by the high-level client which tries to parse the response body as if it was an error, although it most likely was not. Instead of signaling the deprecation warnings clearly, the high-level client ends up breaking with:

```
   > Throwable #1: ElasticsearchStatusException[Unable to parse response body]; nested: ResponseException[method [PUT], host [http://[::1]:33451], URI [/dev_product/product/test.com%2F48923AQ3?timeout=1m], status line [HTTP/1.1 201 Created]
   > Warnings: [[types removal] Specifying types in document index requests is deprecated, use the typeless endpoints instead (/{index}/_doc/{id}, /{index}/_doc, or /{index}/_create/{id}).]
   > ¿f_indexkdev_producte_typegproductc_idqtest.com/48923AQ3h_versionfresultgcreatedg_shards¿etotaljsuccessfulffailedÿg_seq_nom_primary_termÿ]; nested: ResponseException[method [PUT], host [http://[::1]:33451], URI [/dev_product/product/test.com%2F48923AQ3?timeout=1m], status line [HTTP/1.1 201 Created]
   > Warnings: [[types removal] Specifying types in document index requests is deprecated, use the typeless endpoints instead (/{index}/_doc/{id}, /{index}/_doc, or /{index}/_create/{id}).]
   > ¿f_indexkdev_producte_typegproductc_idqtest.com/48923AQ3h_versionfresultgcreatedg_shards¿etotaljsuccessfulffailedÿg_seq_nom_primary_termÿ];
   >    at __randomizedtesting.SeedInfo.seed([CE50F89A4C8ADC37:26EE63FC8DB7EE6D]:0)
   >    at org.elasticsearch.client.RestHighLevelClient.parseResponseException(RestHighLevelClient.java:1681)
   >    at org.elasticsearch.client.RestHighLevelClient.internalPerformRequest(RestHighLevelClient.java:1442)
   >    at org.elasticsearch.client.RestHighLevelClient.performRequest(RestHighLevelClient.java:1399)
   >    at org.elasticsearch.client.RestHighLevelClient.performRequestAndParseEntity(RestHighLevelClient.java:1369)
   >    at org.elasticsearch.client.RestHighLevelClient.index(RestHighLevelClient.java:817)
   >    at org.elasticsearch.client.CrudIT.testWeirdId(CrudIT.java:1379)
   >    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
   >    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
   >    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
   >    at java.base/java.lang.reflect.Method.invoke(Method.java:564)
   >    at java.base/java.lang.Thread.run(Thread.java:844)
   >    Suppressed: ParsingException[Failed to parse object: expecting field with name [error] but found [_index]]
   >            at org.elasticsearch.common.xcontent.XContentParserUtils.ensureFieldName(XContentParserUtils.java:50)
   >            at org.elasticsearch.ElasticsearchException.failureFromXContent(ElasticsearchException.java:587)
   >            at org.elasticsearch.rest.BytesRestResponse.errorFromXContent(BytesRestResponse.java:169)
   >            at org.elasticsearch.client.RestHighLevelClient.parseEntity(RestHighLevelClient.java:1701)
   >            at org.elasticsearch.client.RestHighLevelClient.parseResponseException(RestHighLevelClient.java:1677)
   >            ... 41 more
   > Caused by: org.elasticsearch.client.ResponseException: method [PUT], host [http://[::1]:33451], URI [/dev_product/product/test.com%2F48923AQ3?timeout=1m], status line [HTTP/1.1 201 Created]
   > Warnings: [[types removal] Specifying types in document index requests is deprecated, use the typeless endpoints instead (/{index}/_doc/{id}, /{index}/_doc, or /{index}/_create/{id}).]
   > ¿f_indexkdev_producte_typegproductc_idqtest.com/48923AQ3h_versionfresultgcreatedg_shards¿etotaljsuccessfulffailedÿg_seq_nom_primary_termÿ
   >    at org.elasticsearch.client.RestClient$SyncResponseListener.get(RestClient.java:690)
   >    at org.elasticsearch.client.RestClient.performRequest(RestClient.java:218)
   >    at org.elasticsearch.client.RestHighLevelClient.internalPerformRequest(RestHighLevelClient.java:1429)
   >    ... 40 more
   > Caused by: org.elasticsearch.client.ResponseException: method [PUT], host [http://[::1]:33451], URI [/dev_product/product/test.com%2F48923AQ3?timeout=1m], status line [HTTP/1.1 201 Created]
   > Warnings: [[types removal] Specifying types in document index requests is deprecated, use the typeless endpoints instead (/{index}/_doc/{id}, /{index}/_doc, or /{index}/_create/{id}).]
   > ¿f_indexkdev_producte_typegproductc_idqtest.com/48923AQ3h_versionfresultgcreatedg_shards¿etotaljsuccessfulffailedÿg_seq_nom_primary_termÿ
   >    at org.elasticsearch.client.RestClient$1.completed(RestClient.java:304)
   >    at org.elasticsearch.client.RestClient$1.completed(RestClient.java:294)
   >    at org.apache.http.concurrent.BasicFuture.completed(BasicFuture.java:119)
   >    at org.apache.http.impl.nio.client.DefaultClientExchangeHandlerImpl.responseCompleted(DefaultClientExchangeHandlerImpl.java:177)
   >    at org.apache.http.nio.protocol.HttpAsyncRequestExecutor.processResponse(HttpAsyncRequestExecutor.java:436)
   >    at org.apache.http.nio.protocol.HttpAsyncRequestExecutor.inputReady(HttpAsyncRequestExecutor.java:326)
   >    at org.apache.http.impl.nio.DefaultNHttpClientConnection.consumeInput(DefaultNHttpClientConnection.java:265)
   >    at org.apache.http.impl.nio.client.InternalIODispatch.onInputReady(InternalIODispatch.java:81)
   >    at org.apache.http.impl.nio.client.InternalIODispatch.onInputReady(InternalIODispatch.java:39)
   >    at org.apache.http.impl.nio.reactor.AbstractIODispatch.inputReady(AbstractIODispatch.java:114)
   >    at org.apache.http.impl.nio.reactor.BaseIOReactor.readable(BaseIOReactor.java:162)
   >    at org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvent(AbstractIOReactor.java:337)
   >    at org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvents(AbstractIOReactor.java:315)
   >    at org.apache.http.impl.nio.reactor.AbstractIOReactor.execute(AbstractIOReactor.java:276)
   >    at org.apache.http.impl.nio.reactor.BaseIOReactor.execute(BaseIOReactor.java:104)
   >    at org.apache.http.impl.nio.reactor.AbstractMultiworkerIOReactor$Worker.run(AbstractMultiworkerIOReactor.java:588)
   >    ... 1 more
```

The cause of the failure does contain the deprecation warnings, but it is not so obvious that they did cause the exception to be thrown. The problem here is that a ResponseException gets thrown based on a response that returned OK status code, and no additional exception message is provided to signal what the problem was. The response itself does not necessarily tell the reason of the failure. That same response would not cause any exception with strict deprecation disabled. I believe we should use a different exception for deprecation warnings.

--- a/client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java
+++ b/client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java
@@ -2020,15 +2020,16 @@ public class RestHighLevelClient implements Closeable {
         Response response = responseException.getResponse();
         HttpEntity entity = response.getEntity();
         ElasticsearchStatusException elasticsearchException;
+        RestStatus restStatus = RestStatus.fromCode(response.getStatusLine().getStatusCode());
+
         if (entity == null) {
             elasticsearchException = new ElasticsearchStatusException(
-                    responseException.getMessage(), RestStatus.fromCode(response.getStatusLine().getStatusCode()), responseException);
+                    responseException.getMessage(), restStatus, responseException);
         } else {
             try {
                 elasticsearchException = parseEntity(entity, BytesRestResponse::errorFromXContent);
                 elasticsearchException.addSuppressed(responseException);
             } catch (Exception e) {
-                RestStatus restStatus = RestStatus.fromCode(response.getStatusLine().getStatusCode());
                 elasticsearchException = new ElasticsearchStatusException("Unable to parse response body", restStatus, responseException);
                 elasticsearchException.addSuppressed(e);
             }

--- a/client/rest-high-level/src/test/java/org/elasticsearch/client/MockRestHighLevelTests.java
+++ b/client/rest-high-level/src/test/java/org/elasticsearch/client/MockRestHighLevelTests.java
@@ -1,0 +1,73 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.client;
+
+import org.apache.http.HttpHost;
+import org.apache.http.ProtocolVersion;
+import org.apache.http.RequestLine;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.message.BasicRequestLine;
+import org.apache.http.message.BasicStatusLine;
+import org.elasticsearch.test.ESTestCase;
+import org.junit.Before;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.mockito.Matchers.any;
+import static org.mockito.Mockito.doThrow;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class MockRestHighLevelTests extends ESTestCase {
+    private RestHighLevelClient client;
+    private static final List<String> WARNINGS = Collections.singletonList("Some Warning");
+
+    @Before
+    private void setupClient() throws IOException {
+        final RestClient mockClient = mock(RestClient.class);
+        final Response mockResponse = mock(Response.class);
+
+        when(mockResponse.getHost()).thenReturn(new HttpHost("localhost", 9200));
+        when(mockResponse.getWarnings()).thenReturn(WARNINGS);
+
+        ProtocolVersion protocol = new ProtocolVersion("HTTP", 1, 1);
+        when(mockResponse.getStatusLine()).thenReturn(new BasicStatusLine(protocol, 200, "OK"));
+
+        RequestLine requestLine = new BasicRequestLine(HttpGet.METHOD_NAME, "/_blah", protocol);
+        when(mockResponse.getRequestLine()).thenReturn(requestLine);
+
+        WarningFailureException expectedException = new WarningFailureException(mockResponse);
+        doThrow(expectedException).when(mockClient).performRequest(any());
+
+        client = new RestHighLevelClient(mockClient, RestClient::close, Collections.emptyList());
+    }
+
+    public void testWarningFailure() {
+        WarningFailureException exception = expectThrows(WarningFailureException.class,
+            () -> client.info(RequestOptions.DEFAULT));
+        assertThat(exception.getMessage(), equalTo("method [GET], host [http://localhost:9200], URI [/_blah], " +
+            "status line [HTTP/1.1 200 OK]"));
+        assertNull(exception.getCause());
+        assertThat(exception.getResponse().getWarnings(), equalTo(WARNINGS));
+    }
+}

--- a/client/rest/src/main/java/org/elasticsearch/client/ResponseException.java
+++ b/client/rest/src/main/java/org/elasticsearch/client/ResponseException.java
@@ -32,7 +32,7 @@ import java.util.Locale;
  */
 public final class ResponseException extends IOException {
 
-    private Response response;
+    private final Response response;
 
     public ResponseException(Response response) throws IOException {
         super(buildMessage(response));
@@ -49,7 +49,7 @@ public final class ResponseException extends IOException {
         this.response = e.getResponse();
     }
 
-    private static String buildMessage(Response response) throws IOException {
+    static String buildMessage(Response response) throws IOException {
         String message = String.format(Locale.ROOT,
             "method [%s], host [%s], URI [%s], status line [%s]",
             response.getRequestLine().getMethod(),

--- a/client/rest/src/main/java/org/elasticsearch/client/RestClient.java
+++ b/client/rest/src/main/java/org/elasticsearch/client/RestClient.java
@@ -540,7 +540,7 @@ public class RestClient implements Closeable {
                     if (isSuccessfulResponse(statusCode) || ignoreErrorCodes.contains(response.getStatusLine().getStatusCode())) {
                         onResponse(node);
                         if (thisWarningsHandler.warningsShouldFailRequest(response.getWarnings())) {
-                            listener.onDefinitiveFailure(new ResponseException(response));
+                            listener.onDefinitiveFailure(new WarningFailureException(response));
                         } else {
                             listener.onSuccess(response);
                         }
@@ -925,6 +925,9 @@ public class RestClient implements Closeable {
                  * like the asynchronous API. We wrap the exception so that the caller's
                  * signature shows up in any exception we throw.
                  */
+                if (exception instanceof WarningFailureException) {
+                    throw new WarningFailureException((WarningFailureException) exception);
+                }
                 if (exception instanceof ResponseException) {
                     throw new ResponseException((ResponseException) exception);
                 }

--- a/client/rest/src/main/java/org/elasticsearch/client/WarningFailureException.java
+++ b/client/rest/src/main/java/org/elasticsearch/client/WarningFailureException.java
@@ -1,0 +1,58 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.client;
+
+import java.io.IOException;
+
+import static org.elasticsearch.client.ResponseException.buildMessage;
+
+/**
+ * This exception is used to indicate that one or more {@link Response#getWarnings()} exist
+ * and is typically used when the {@link RestClient} is set to fail by setting
+ * {@link RestClientBuilder#setStrictDeprecationMode(boolean)} to `true`.
+ */
+// This class extends RuntimeException in order to deal with wrapping that is done in FutureUtils on exception.
+// if the exception is not of type ElasticsearchException or RuntimeException it will be wrapped in a UncategorizedExecutionException
+public final class WarningFailureException extends RuntimeException {
+
+    private final Response response;
+
+    public WarningFailureException(Response response) throws IOException {
+        super(buildMessage(response));
+        this.response = response;
+    }
+
+    /**
+     * Wrap a {@linkplain WarningFailureException} with another one with the current
+     * stack trace. This is used during synchronous calls so that the caller
+     * ends up in the stack trace of the exception thrown.
+     */
+    WarningFailureException(WarningFailureException e) {
+        super(e.getMessage(), e);
+        this.response = e.getResponse();
+    }
+
+    /**
+     * Returns the {@link Response} that caused this exception to be thrown.
+     */
+    public Response getResponse() {
+        return response;
+    }
+}

--- a/client/rest/src/test/java/org/elasticsearch/client/RestClientSingleHostTests.java
+++ b/client/rest/src/test/java/org/elasticsearch/client/RestClientSingleHostTests.java
@@ -467,9 +467,9 @@ public class RestClientSingleHostTests extends RestClientTestCase {
         if (expectFailure) {
             try {
                 restClient.performRequest(request);
-                fail("expected ResponseException from warnings");
+                fail("expected WarningFailureException from warnings");
                 return;
-            } catch (ResponseException e) {
+            } catch (WarningFailureException e) {
                 if (false == warningBodyTexts.isEmpty()) {
                     assertThat(e.getMessage(), containsString("\nWarnings: " + warningBodyTexts));
                 }

